### DISCUSSION TITLE
How does goja JSVM exception throwing work?

### DISCUSSION THREAD

**Discussion by dschissler** [pocketbase/pocketbase#3899] [Q&A]
I've made significant progress towards `caspoc` (Casbin Pocketbase) integration.  I'm using it already, it's just that I haven't worked on it for JSVM yet.  I'm conceptually making it so that it's good for both Golang and JSVM, but for the moment I'm only simulating the JSVM code in my head until I get a good API.

The second parameter of `.middleware` is a function with the following signature:

```golang
type SubjectResolver func(*models.RequestInfo) string
```

My question is about errors and exceptions.  Perhaps I could remove the `try catch` to make any errors be handled in the level above.  My question is if goja is smart enough to automatically return an unhandled exception as a Golang error.

So if I were to change the signature to the following, then would goja automatically return the unhandled exception in the form of error and with the string as its zero value?:

```golang
type SubjectResolver func(*models.RequestInfo) (string, error)
```

```javascript
const $caspoc = $app.cache().get("caspoc")

const middleware = $caspoc(
  `${__hooks}/model.config`,
  `${__hooks}/policy.csv`,
).middleware({
  admin: "root",
  authRecordUnknown: "unknown",
  authRecordUnverified: "unverified",
}, ({authRecord}) => {
  try {
    return $app
      .dao()
      .findRecordById(
        models.UserGroups,
        authRecord.getString("group"),
      )
      .getString("name")
  } catch {
    return ""
  }
})

routerAdd("GET", "/hello/restrict", (c) => {
  return c.json(200, { "message": "Hello Restrict" })
}, middleware.Restrict())

routerAdd("GET", "/hello/redirect", (c) => {
  return c.json(200, { "message": "Hello Restrict" })
}, middleware.Redirect(
  "/allowed-redirect-path",
  "/fail-fallback-redirect-path",
  {
    "normal": "/fail-redirect-path-for-normal-group",
    "premium": "/fail-redirect-path-for-premium-group",
  },
))

routerAdd("GET", "/hello/restrict-with-redirect", (c) => {
  return c.json(200, { "message": "Hello Restrict" })
}, middleware.RestrictWithRedirect(
  "/fail-fallback-path",
  {
    "normal": "/upgrade-to-premium-plus",
    "premium": "/upgrade-to-premium-plus-platinum-package",
  },
))
```

**Comment by ganigeorgiev** [comment]
The last return value from a function/method is skipped from the return result and if non-empty will be converted automatically to an exception. In other words, the behavior should be something like:

```
func Example0() error
- on nil error     -> return undefined
- on non-nil error -> throw exception

func Example1() (string, error)
- on nil error     -> return string
- on non-nil error -> throw exception

func Example2() (int, string, error)
- on nil error     -> return [int, string]
- on non-nil error -> throw exception
```

(_see also the [goja godoc](https://pkg.go.dev/github.com/dop251/goja#hdr-Nil)_)

**Reply by dschissler** [reply]
I want to turn this:

```golang
type SubjectResolver func(*models.RequestInfo) string
```

```javascript
}, ({authRecord}) => {
  try {
    return $app
      .dao()
      .findRecordById(
        models.UserGroups,
        authRecord.getString("group"),
      )
      .getString("name")
  } catch {
    return ""
  }
}
```

Into this:

```golang
type SubjectResolver func(*models.RequestInfo) (string, error)
```

```javascript
}, ({authRecord}) => {
  return $app
    .dao()
    .findRecordById(
      models.UserGroups,
      authRecord.getString("group"),
    )
    .getString("name")
})
```

Will I automatically get back (empty string, non-nil error value) upon `findRecordById` throwing an exception (when it returns a non-nil error)?

**Reply by dschissler** [reply]
I understand that, but I'm confused about how an unhandled thrown exception in a goja function gets converted into function results back in Golang, if Golang was calling that function.  Does goja look for the first error return value to stuff the exception into?

**Reply by ganigeorgiev** [reply]
I'm not sure I follow.

If you have a Go function that returns 2 values - string and error, the JS translated method will will throw an exception if the Go error is non-nil.

**Reply by dschissler** [reply]
If the function signature defined in Golang is `type SubjectResolver func(*models.RequestInfo) (string, error)`, then will this function executed in goja that throws an exception cause the return value to be (empty string, non-nil error)?

```javascript
({authRecord}) => {
  throw "will this exception give me (empty string, non-nil error) back in Golang land?"

  return $app
    .dao()
    .findRecordById(
      models.UserGroups,
      authRecord.getString("group"),
    )
    .getString("name")
}
```

I would play with this directly instead of asking you, but that custom goja setup of mine is running on a Raspberry Pi, and it's a private project that I don't have time for now.  I'm putting out a bit of extra time to make the API goja friendly, but it's all kind of new to me, and so I'm not setup on all environments, and compiling Golang code on the Raspberry Pi just for R&D is crappy.

**Reply by ganigeorgiev** [reply]
Do you mean that you have somewhere in your Go code that accept `SubjectResolver` function and you want to pass such function from the JS?

In that case you don't have to worry about the JS return types. The goja wrapped function would attempt to automatically convert the JS function result to its reflected Go equivalent. Or in other words - if you throw an exception from the JS, the above `SubjectResolver` Go function will receive  empty string and the exception as error values.

But keep in mind that passing functions from JS -> Go may not work if that function will be later executed in a different goroutine.

**Reply by dschissler** [reply]
Thanks for clarifying that

> But keep in mind that passing functions from JS -> Go may not work if that function will be later executed in a different goroutine.

It's a route middleware.  Aren't http routes excecuted on their own goroutine from a pool already?  I was under the impression that it's already concurrent and that there are things that are generally understood about how not to screw up with that.

Are you referring to the JS functions being serialized?  I'm not able to reason about how any of that works.

**Reply by ganigeorgiev** [reply]
It depends where the JS function is registered and called.
If it is defined inside a route, hook, middleware, etc. handler, then it is safe to be called within that handler because it runs in its own goja runtime.
But if your JS function is registered in goroutine A and you attempt to call it in goroutine B, then it could panic in case of a concurrent access because the function will try to alter the runtime state that may have changed since it was registered, aka. you may end up in a situation like my initial goja integration from https://github.com/pocketbase/pocketbase/discussions/2825.

**Reply by dschissler** [reply]
Thanks.  I think that I understand.  I'll review that entire subject matter.  At the least I'll learn some more details that might make me look smart one day.

**Comment by dschissler** [comment]
I've made a core `Auth` that the Builder can make to reuse across many middlewares.

```golang
// auth.go

package caspoc

import (
	"github.com/casbin/casbin/v2"
	"github.com/labstack/echo/v5"
	"github.com/pocketbase/pocketbase/apis"
	"github.com/pocketbase/pocketbase/models"
)

type SubjectResolver func(*models.RequestInfo) string

func NewAuth(
	enforcer *casbin.Enforcer,
	subjectResolver SubjectResolver,
) *Auth {
	return &Auth{
		enforcer:        enforcer,
		subjectResolver: subjectResolver,
	}
}

type Auth struct {
	enforcer        *casbin.Enforcer
	subjectResolver SubjectResolver
}

func (a *Auth) AllowRoute(c echo.Context) (bool, string) {
	s := a.subjectResolver(apis.RequestInfo(c))
	if s == "" {
		return false, s
	}

	sub := "@" + s
	obj := c.Path()
	act := c.Request().Method

	alw, err := a.enforcer.Enforce(sub, obj, act)
	return alw && err == nil, s
}
```

StandardResolver is a `*models.RequestInfo` mapper to a casbin subject (which is most likely to be a user group name).  That is why question queries are needed.  I left that part as flexible as possible because there are massive numbers of ways to organize that relationship.  This requires a custom query.

The StandardResolver can optionally pick up on PB admin tokens, uknown/anonymous access, and unverified by just setting the name of that group in the options struct.  If you don't anything in StandardOptions then it basically comes over almost the same as just writing your own `caspoc.SubjectResolver`.  It's barely anymore work to do it, but probably almost everyone could use the standard one.

```golang
// resolver.go

package caspoc

import (
	"github.com/pocketbase/pocketbase/models"
)

type StandardOptions struct {
	Admin                string
	AuthRecordUnknown    string
	AuthRecordUnverified string
}

func StandardResolver(
	options StandardOptions,
	subjectResolver SubjectResolver,
) SubjectResolver {
	useAdmin := options.Admin != ""
	useUnverified := options.AuthRecordUnverified != ""

	return func(reqInfo *models.RequestInfo) string {
		if useAdmin && reqInfo.Admin != nil {
			return options.Admin
		}

		user := reqInfo.AuthRecord
		if user == nil {
			return options.AuthRecordUnknown
		} else if useUnverified && !user.Verified() {
			return options.AuthRecordUnverified
		}

		sub := subjectResolver(reqInfo)
		if sub == "" {
			return options.AuthRecordUnknown
		}
		return sub
	}
}
```

This is just a taste.  I think that I've managed to achieve my goals already.  Now it's just a bit of polish before I have to set this down.  I'll be delivering it as a code dump that people can copy into their projects.
